libsplinter/src/admin/service/shared.rs

```diff
@@ -841,46 +929,131 @@ impl AdminServiceShared {
             for peered_payload in fully_peered {
-                match peered_payload.payload_type {
-                    PayloadType::Circuit(payload) => {
-                        self.pending_circuit_payloads.push_back(payload)
+                self.pending_protocol_payloads.push(peered_payload);
+            }
+
+            // Ignore own admin service
+            if peer_id == admin_service_id(self.node_id()) {
+                return Ok(());
+            }
+
+            if let Some(ref network_sender) = self.network_sender {
+                debug!(
+                    "Sending service protocol request to {}",
+                    admin_service_id(peer_id)
+                );
+                let mut request = ServiceProtocolVersionRequest::new();
+                request.set_protocol_min(ADMIN_SERVICE_PROTOCOL_MIN);
+                request.set_protocol_max(ADMIN_PROTOCOL_VERSION);
+                let mut msg = AdminMessage::new();
+                msg.set_message_type(AdminMessage_Type::SERVICE_PROTOCOL_VERSION_REQUEST);
+                msg.set_protocol_request(request);
+
+                let envelope_bytes = msg.write_to_bytes().map_err(|err| {
+                    AuthorizationCallbackError(format!(
+                        "Unable to send service protocol request: {}",
+                        err
+                    ))
+                })?;
+
+                network_sender
+                    .send(&admin_service_id(peer_id), &envelope_bytes)
+                    .map_err(|err| {
+                        AuthorizationCallbackError(format!(
+                            "Unable to send service protocol request: {}",
+                            err
+                        ))
+                    })?;
+            } else {
+                return Err(AuthorizationCallbackError(format!(
+                    "AdminService is not started {}",
+                    peer_id
+                )));
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn on_protocol_agreement(
+        &mut self,
+        service_id: &str,
+        protocol: u32,
+    ) -> Result<(), AdminSharedError> {
+        let mut pending_protocol_payloads =
+            std::mem::replace(&mut self.pending_protocol_payloads, vec![]);
+        for pending_protocol_payload in pending_protocol_payloads.iter_mut() {
+            match protocol {
+                0 => {
+                    if pending_protocol_payload
+                        .missing_protocol_ids
+                        .iter()
+                        .any(|missing_protocol_id| missing_protocol_id == service_id)
+                    {
+                        warn!(
+                            "Dropping circuit request including service {}, \
+                             due to protocol mismatch",
+                            service_id
+                        );
+                        pending_protocol_payload.missing_protocol_ids.clear();
```

Comment:
Doesn't this delete all of the missing protocol ids?

Reply:
yes